### PR TITLE
LoveLace Light Card

### PR DESCRIPTION
**Pull request by zsarnett**
I put this together while not knowing what to do with a websocket call

image

image

## Configuration Variables

| Name | Type | Default | Description
| ---- | ---- | ------- | -----------
| type | string | **Required** | `light`
| entity | string | optional | Required `light.my_light`
| name | string | friendly_name | Name of card

## Example

```yaml
- type: light
  entity: light.living_room
```

```yaml
- type: entity-button
  entity: light.bedroom
  name: Living Room
```

### CODE REVIEW

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+export class HuiLightCard extends hassLocalizeLitMixin(LitElement)
+  implements LovelaceCard {
+  public hass?: HomeAssistant;
+  protected config?: Config;
+
+  static get properties() {
+    return {
+      hass: {},
+      config: {},
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-light-card.ts]: Make this `_config`, it's a private property.

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+interface Config extends LovelaceConfig {
+  entity: string;
+  name?: string;
+}
+
+export class HuiLightCard extends hassLocalizeLitMixin(LitElement)
+  implements LovelaceCard {
+  public hass?: HomeAssistant;
+  protected config?: Config;
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-light-card.ts]: private.

**Comment by balloob** [comment]
What happens if light has no brightness?

**Comment by zsarnett** [comment]
> What happens if light supports no brightness?

If it has no brightness its set to 0% or off. The only time it should have no brightness. If the entity they are using has no brightness use the button card :)

**Comment by zsarnett** [comment]
Made the Brightness Percentage above Name so that a longer name can be added.

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+      }
+      .brightness {
+        1.2rem;
+      }
+    </style>
+    `;
+  }
+
+  private _dragEvent(e) {
+    this.shadowRoot!.querySelector(".brightness")!.innerHTML = e.value + "%";
```
Comment (zsarnett) [inline src/panels/lovelace/cards/hui-light-card.ts]: `TypeError: Cannot read property 'nodeType' of null` 

I get the above error when dragging the brightness

Comment (balloob) [reply src/panels/lovelace/cards/hui-light-card.ts]: It means `.brightness` cannot be found inside the shadow root.

Comment (balloob) [reply src/panels/lovelace/cards/hui-light-card.ts]: Or `this.shadowRoot` is null, which seems odd.

Comment (zsarnett) [reply src/panels/lovelace/cards/hui-light-card.ts]: Actually, I miss spoke. I when I let go to change change the brightness. After something changes The icon doesnt change brightness and I get that error

**Comment by balloob** [comment]
Can you add a demo

**Comment by balloob** [comment]
https://deploy-preview-1874--home-assistant-lovelace-gallery.netlify.com/#demo-hui-light-card

When I click on the slider rail (not the slider knob), it toggles the light. That's not what I would expect. The toggle area of the icon should be limited to the icon only.

The slider knob is not centered on the rail:
image

I don't think that we should always show percentage. Maybe we can only show it when dragging as an overlay on the icon?

File: gallery/src/demos/demo-hui-light-card.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    `;
+  }
+
+  static get properties() {
+    return {
+      _configs: {
+        type: Object,
+        value: CONFIGS,
+      },
+      hass: Object,
```
Comment (balloob) [inline gallery/src/demos/demo-hui-light-card.js]: why did you add this?

File: gallery/src/demos/demo-hui-light-card.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+- type: light
+  entity: light.bed_light
+    `,
+  },
+];
+
+class DemoLightEntity extends PolymerElement {
+  static get template() {
+    return html`
+      <demo-cards id='demos' hass='[[hass]]' configs="[[_configs]]"></demo-cards>
```
Comment (balloob) [inline gallery/src/demos/demo-hui-light-card.js]: Don't add `hass`

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+      ha-icon {
+        display: flex;
+        margin: auto;
+        width: 40%;
+        height: 40%;
+        color: var(--paper-item-icon-color, #44739e);
+        cursor: pointer;
+      }
+      ha-icon[data-domain=light][data-state=on],
+      ha-icon[data-domain=switch][data-state=on],
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-light-card.ts]: what

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+
+    return html`
+      ${this.renderStyle()}
+      <ha-card>
+        <div id="root">
+          <div id="light"></div>
+          <div id="tooltip">
+            <div class="icon-state">
+              <ha-icon
+                data-domain="${computeStateDomain(stateObj)}"
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-light-card.ts]: remove?

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+    return html`
+    ${roundSliderStyle}
+    <style>
+      :host {
+        display: block;
+      }
+      ha-card {
+        overflow: hidden;
+      }
+      #root {
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-light-card.ts]: Why is this it's own div and not applied to `ha-card` ?

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+  }
+
+  protected shouldUpdate(changedProps) {
+    if (changedProps.get("hass")) {
+      return (
+        changedProps.get("hass").states[this._config!.entity] !==
+        this.hass!.states[this._config!.entity]
+      );
+    }
+    if (changedProps.get("_config")) {
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-light-card.ts]: No need to check for this.

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+              <div class="name">${this._config.name ||
+                computeStateName(stateObj)}</div>
+            </div>
+          </div>
+        </div>
+      </ha-card>
+    `;
+  }
+
+  protected shouldUpdate(changedProps) {
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-light-card.ts]: Add type

File: src/panels/lovelace/cards/hui-light-card.ts (hunk 9 above)
Comment (balloob) [reply src/panels/lovelace/cards/hui-light-card.ts]: If config was updated, we re-render.

Comment (zsarnett) [reply src/panels/lovelace/cards/hui-light-card.ts]: I thought you said to add this check in the other PR 🤔

File: src/panels/lovelace/cards/hui-light-card.ts (hunk 6 above)
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-light-card.ts]: Ah yea I copied the button card icon styling. oops

File: gallery/src/demos/demo-hui-light-card.js (hunk 4 above)
Comment (zsarnett) [reply gallery/src/demos/demo-hui-light-card.js]: uhh Its how the button card is in the demo. Ill remove

**Comment by zsarnett** [comment]
I can not replicate the clicking on the rail toggling the light effect

File: src/panels/lovelace/cards/hui-light-card.ts (hunk 9 above)
Comment (balloob) [reply src/panels/lovelace/cards/hui-light-card.ts]: Not for `_config`, just for the state objects.

**Comment by balloob** [comment]
The middle of the icon is not clickable because the brightness label still exists when not in use:

image

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+        color: var(--paper-item-icon-active-color, #FDD835);
+      }
+      ha-icon[data-state=unavailable] {
+        color: var(--state-icon-unavailable-color);
+      }
+      .name {
+        padding-top: 40px;
+        font-size: 1.2rem;
+      }
+      .brightness {
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-light-card.ts]: The text is not always very visible as the icon color, which is the background of brightness, changes.

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+
+  private _showBrightness() {
+    clearTimeout(brightnessTimout);
+    this.shadowRoot!.querySelector(".brightness")!.classList.add(
+      "show_brightness"
+    );
+  }
+
+  private _hideBrightness() {
+    brightnessTimout = setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-light-card.ts]: Nope. Meant to make a comment on the best way to do this. This is how it's done for the notification drawer but there is one of those.

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+            <ha-icon
+              data-state="${stateObj.state}"
+              .icon="${stateIcon(stateObj)}"
+              style="${styleMap({
+                filter: this._computeBrightness(stateObj),
+                color: this._computeColor(stateObj),
+              })}"
+              @click="${this._handleClick}"
+            ></ha-icon>
+            <div class="brightness">%</div>
```
[the remark this answers is not in the document]
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-light-card.ts]: Oops good catch

**Comment by zsarnett** [comment]
@balloob clicking the slider rail works in the demo of round slider because they aren't using a custom overlay for the tool tip. Because we are the overlay is above the rail and can't make it a higher z index with out overlaying a white background on our tool tip. At least when I try to change the zindex of the rail that's what I get

File: src/panels/lovelace/cards/hui-light-card.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (balloob) [reply src/panels/lovelace/cards/hui-light-card.ts]: what about an instance variable? `this._brightnessTimeout`

[the remark this answers is not in the document]
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-light-card.ts]: I'm always so conflicted when creating properties in LitElement. I would need to set that as ` private _brightneesTimeout?: Any;`

**Comment by frenck** [comment]
Just a small note: Could not find a related PR on the documentation repository.

File: src/panels/lovelace/cards/hui-light-card.ts (hunk 11 above)
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-light-card.ts]: Added a stroke to it but not sure how material design esque this is

File: src/panels/lovelace/cards/hui-light-card.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (balloob) [reply src/panels/lovelace/cards/hui-light-card.ts]: Never use `any`. Hover with your mouse over `setTimeout`, what is the defined returned value? 

With TS you have to define the type like you said, then just assign it wherever you want yes.

File: src/panels/lovelace/cards/hui-light-card.ts (hunk 11 above)
Comment (balloob) [reply src/panels/lovelace/cards/hui-light-card.ts]: it's a good start, we can iterate on it.

File: script/version_bump.js [hunk 14]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env node
```
Comment (balloob) [inline script/version_bump.js]: Nice script.

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 15]
```diff
@@ -0,0 +1,313 @@
...
+interface Config extends LovelaceConfig {
+  entity: string;
+  name?: string;
+}
+
+export class HuiLightCard extends hassLocalizeLitMixin(LitElement)
+  implements LovelaceCard {
+  public hass?: HomeAssistant;
+  private _config?: Config;
+  private brightnessTimout?: any;
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-light-card.ts]: let's prefix names of private vars with `_`. I know it's "double", but we're also interacting with JS and they don't know the concept of private.

Comment (balloob) [reply src/panels/lovelace/cards/hui-light-card.ts]: If you don't know the type, you can hover with your mouse on the function (if it's a built-in), otherwise use `unknown`. With `unknown`, you are not allowed to do any operation besides passing it on or comparing it, until you know the type. With `any`, anything goes and we're back to caveman untyped world.

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 16]
```diff
@@ -0,0 +1,313 @@
...
+      this.shadowRoot!.querySelector(".brightness")!.classList.remove(
+        "show_brightness"
+      );
+    }, 500);
+  }
+
+  private _setBrightness(e) {
+    this.hass!.callService("light", "turn_on", {
+      entity_id: this._config!.entity,
+      brightness_pct: e.value,
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-light-card.ts]: Now I know why the gallery is flaky. We don't have support for `brightness_pct` in our demo hass for the gallery.

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 17]
```diff
@@ -0,0 +1,313 @@
...
+    }
+    const brightness = stateObj.attributes.brightness;
+    return `brightness(${(brightness + 245) / 5}%)`;
+  }
+
+  private _computeColor(stateObj) {
+    if (!stateObj.attributes.hs_color) {
+      return "";
+    }
+    const hue = stateObj.attributes.hs_color[0];
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-light-card.ts]: Array destructuring is a thing in JS and TS:

```ts
const [hue, sat] = stateObj.attributes.hs_color;
```

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+
+  private _computeBrightness(stateObj) {
+    if (!stateObj.attributes.brightness) {
+      return "";
+    }
+    const brightness = stateObj.attributes.brightness;
+    return `brightness(${(brightness + 245) / 5}%)`;
+  }
+
+  private _computeColor(stateObj) {
```
[the remark this answers is not in the document]
Comment (balloob) [reply src/panels/lovelace/cards/hui-light-card.ts]: marked as resolved but this state object does not have a type yet. Add types to all functions you define in your class.

File: src/panels/lovelace/cards/hui-light-card.ts (hunk 16 above)
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-light-card.ts]: How can I add that?

File: src/panels/lovelace/cards/hui-light-card.ts (hunk 15 above)
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-light-card.ts]: Gotcha. I started to do that that but it was a NodeJS Type so I left it off. I added it in because it was accepted with no errors

File: src/panels/lovelace/cards/hui-light-card.ts (hunk 16 above)
Comment (balloob) [reply src/panels/lovelace/cards/hui-light-card.ts]: Where did you look?

Comment (balloob) [reply src/panels/lovelace/cards/hui-light-card.ts]: You need to update the logic here https://github.com/home-assistant/home-assistant-polymer/blob/dev/gallery/src/data/entity.js#L56-L65

File: src/panels/lovelace/cards/hui-light-card.ts [hunk 19]
```diff
@@ -0,0 +1,312 @@
...
+interface Config extends LovelaceConfig {
+  entity: string;
+  name?: string;
+}
+
+export class HuiLightCard extends hassLocalizeLitMixin(LitElement)
+  implements LovelaceCard {
+  public hass?: HomeAssistant;
+  private _config?: Config;
+  private _brightnessTimout?: NodeJS.Timer;
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-light-card.ts]: We're running in a browser, not under NodeJS. Also:

image

`number`

**Review by balloob — APPROVED** [review]
ok to merge when fixing type of setTimeout.

File: src/panels/lovelace/cards/hui-light-card.ts (hunk 19 above)
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-light-card.ts]: When changing to `number` I am getting `Type Time is not assignable to Type Number` here: `this._brightnessTimout = setTimeout(() => {`

I think Node is messing with things or something.. Because researching setTimeout shows it returns a "unique" number that is used when clearing the timeout

Comment (zsarnett) [reply src/panels/lovelace/cards/hui-light-card.ts]: Fix this by using window.setTimeout so node doesnt overload
